File: agentprimordia/internal/agent/hooks/hooks.go
// Package hooks 提供 Agent 钩子管理器
package hooks

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"sync/atomic"
	"time"

	"agentprimordia/internal/agent/core"
)

// hookContextPool 复用 HookContext 实例。
var hookContextPool = sync.Pool{
	New: func() any {
		return &HookContext{}
	},
}

// AcquireHookContext 从池中获取一个已重置的 HookContext。
func AcquireHookContext() *HookContext {
	hctx := hookContextPool.Get().(*HookContext)
	hctx.Reset()
	return hctx
}

// ReleaseHookContext 归还 HookContext 到池中。
func ReleaseHookContext(hctx *HookContext) {
	if hctx == nil {
		return
	}
	hctx.Reset()
	hookContextPool.Put(hctx)
}

// Reset 清空 HookContext 的所有字段。
func (h *HookContext) Reset() {
	if h == nil {
		return
	}
	h.AgentID = ""
	h.RequestID = ""
	h.SessionID = ""
	h.Point = ""
	h.Turn = 0
	h.Message = nil
	h.Response = nil
	h.ToolCall = nil
	h.ToolResult = nil
	h.Error = nil
	if h.Metadata != nil {
		for k := range h.Metadata {
			delete(h.Metadata, k)
		}
	}
	h.StreamChunk = nil
	h.Duration = 0
	h.OldState = ""
	h.NewState = ""
	h.Reason = ""
	h.MemoryQuery = ""
	h.MemoryResult = nil
	h.ContextWindowUsage = 0
	h.ContextWindowLimit = 0
}

// HookPoint 钩子触发点
type HookPoint string

// 钩子点常量
const (
	HookBeforeRun           HookPoint = "before_run"
	HookAfterRun            HookPoint = "after_run"
	HookBeforeTurn          HookPoint = "before_turn"
	HookAfterTurn           HookPoint = "after_turn"
	HookBeforeLLM           HookPoint = "before_llm"
	HookAfterLLM            HookPoint = "after_llm"
	HookBeforeTool          HookPoint = "before_tool"
	HookAfterTool           HookPoint = "after_tool"
	HookOnError             HookPoint = "on_error"
	HookOnComplete          HookPoint = "on_complete"
	HookBeforeRAG           HookPoint = "before_rag"
	HookAfterRAG            HookPoint = "after_rag"
	HookBeforePipelineStep  HookPoint = "before_pipeline_step"
	HookAfterPipelineStep   HookPoint = "after_pipeline_step"
	HookBeforeHandoff       HookPoint = "before_handoff"
	HookAfterHandoff        HookPoint = "after_handoff"
	HookBeforeParallelAgent HookPoint = "before_parallel_agent"
	HookAfterParallelAgent  HookPoint = "after_parallel_agent"
	HookBeforeDAGNode       HookPoint = "before_dag_node"
	HookAfterDAGNode        HookPoint = "after_dag_node"
	HookOnStream            HookPoint = "on_stream"
	HookOnStreamStart       HookPoint = "on_stream_start"
	HookOnStreamEnd         HookPoint = "on_stream_end"
	HookBeforeMemoryRead    HookPoint = "before_memory_read"
	HookAfterMemoryRead     HookPoint = "after_memory_read"
	HookBeforeMemoryWrite   HookPoint = "before_memory_write"
	HookAfterMemoryWrite    HookPoint = "after_memory_write"
	HookContextWindowUpdate HookPoint = "context_window_update"
	HookContextWindowFull   HookPoint = "context_window_full"
	HookBeforeToolParse     HookPoint = "before_tool_parse"
	HookAfterToolParse      HookPoint = "after_tool_parse"
	HookOnMetricsCollect    HookPoint = "on_metrics_collect"
	HookBeforeShutdown      HookPoint = "before_shutdown"
	HookAfterShutdown       HookPoint = "after_shutdown"
	HookOnStateChange       HookPoint = "on_state_change"
)

// HookContext 钩子执行上下文
type HookContext struct {
	AgentID    string
	RequestID  string
	SessionID  string
	Point      HookPoint
	Turn       int
	Message    *core.Message
	Response   *core.Response
	ToolCall   *core.ToolCall
	ToolResult *core.ToolResult
	Error      error
	Metadata   map[string]any

	StreamChunk *core.StreamEvent
	Duration    time.Duration

	OldState string
	NewState string
	Reason   string

	MemoryQuery  string
	MemoryResult any

	ContextWindowUsage float64
	ContextWindowLimit int
}

// HookFunc 钩子函数类型
type HookFunc func(ctx context.Context, hctx *HookContext) error

// Hooks 是 HookManager 指针的类型别名
type Hooks = *HookManager

// HookPhase 钩子执行阶段
type HookPhase int

const (
	PhaseValidation HookPhase = iota
	PhasePreProcessing
	PhaseExecution
	PhasePostProcessing
)

// Hook 钩子定义
type Hook struct {
	Point     HookPoint
	Func      HookFunc
	Priority  int
	Condition HookCondition
	ID        string
	Phase     HookPhase
}

// HookCondition 钩子条件
type HookCondition func(ctx context.Context, hctx *HookContext) bool

// Always 总是返回 true
func Always(_ context.Context, _ *HookContext) bool { return true }

// OnTurn 在指定 turn 号时触发
func OnTurn(turnNum int) HookCondition {
	return func(_ context.Context, hctx *HookContext) bool {
		return hctx.Turn == turnNum
	}
}

// OnTurnsGreater 在 turn 大于指定值时触发
func OnTurnsGreater(threshold int) HookCondition {
	return func(_ context.Context, hctx *HookContext) bool {
		return hctx.Turn > threshold
	}
}

// OnError 当有错误时触发
func OnError() HookCondition {
	return func(_ context.Context, hctx *HookContext) bool {
		return hctx.Error != nil
	}
}

// OnMetadataKey 当 Metadata 包含指定 key 时触发
func OnMetadataKey(key string) HookCondition {
	return func(_ context.Context, hctx *HookContext) bool {
		if hctx.Metadata == nil {
			return false
		}
		_, ok := hctx.Metadata[key]
		return ok
	}
}

// OnStateTransition 指定状态转换时触发
func OnStateTransition(from, to string) HookCondition {
	return func(_ context.Context, hctx *HookContext) bool {
		return hctx.OldState == from && hctx.NewState == to
	}
}

const hookPointCount = 64

// HookStats 钩子执行统计
type HookStats struct {
	TotalFired   int64
	TotalErrors  int64
	ByPoint      []atomic.Int64
	ByPointError []atomic.Int64
	nameIndex    map[HookPoint]int
}

// NewHookStats 创建钩子执行统计器
func NewHookStats() *HookStats {
	idx := make(map[HookPoint]int, hookPointCount)
	all := []HookPoint{
		HookBeforeRun, HookAfterRun, HookBeforeTurn, HookAfterTurn,
		HookBeforeLLM, HookAfterLLM, HookBeforeTool, HookAfterTool,
		HookOnError, HookOnComplete, HookBeforeRAG, HookAfterRAG,
		HookBeforePipelineStep, HookAfterPipelineStep,
		HookBeforeHandoff, HookAfterHandoff,
		HookBeforeParallelAgent, HookAfterParallelAgent,
		HookBeforeDAGNode, HookAfterDAGNode,
		HookOnStream, HookOnStreamStart, HookOnStreamEnd,
		HookBeforeMemoryRead, HookAfterMemoryRead, HookBeforeMemoryWrite, HookAfterMemoryWrite,
		HookContextWindowUpdate, HookContextWindowFull,
		HookBeforeToolParse, HookAfterToolParse,
		HookOnMetricsCollect, HookBeforeShutdown, HookAfterShutdown, HookOnStateChange,
	}
	for i, p := range all {
		idx[p] = i
	}
	return &HookStats{
		ByPoint:      make([]atomic.Int64, hookPointCount),
		ByPointError: make([]atomic.Int64, hookPointCount),
		nameIndex:    idx,
	}
}

func (s *HookStats) Record(point HookPoint, err error) {
	atomic.AddInt64(&s.TotalFired, 1)
	if idx, ok := s.nameIndex[point]; ok && idx < hookPointCount {
		s.ByPoint[idx].Add(1)
		if err != nil {
			atomic.AddInt64(&s.TotalErrors, 1)
			s.ByPointError[idx].Add(1)
		}
	}
}

func (s *HookStats) Snapshot() map[string]interface{} {
	result := map[string]interface{}{
		"total_fired":  atomic.LoadInt64(&s.TotalFired),
		"total_errors": atomic.LoadInt64(&s.TotalErrors),
		"by_point":     make(map[string]int64, len(s.nameIndex)),
		"by_errors":    make(map[string]int64, len(s.nameIndex)),
	}
	for p, idx := range s.nameIndex {
		if v := s.ByPoint[idx].Load(); v > 0 {
			result["by_point"].(map[string]int64)[string(p)] = v
		}
		if v := s.ByPointError[idx].Load(); v > 0 {
			result["by_errors"].(map[string]int64)[string(p)] = v
		}
	}
	return result
}

// HookManager 钩子管理器
type HookManager struct {
	mu             sync.RWMutex
	hooks          map[HookPoint][]Hook
	stats          *HookStats
	middleware     []HookMiddleware
	hooksSnap      atomic.Pointer[map[HookPoint][]Hook]
	middlewareSnap atomic.Pointer[[]HookMiddleware]
}

// HookMiddleware 中间件接口
type HookMiddleware interface {
	Before(ctx context.Context, hctx *HookContext) error
	After(ctx context.Context, hctx *HookContext, err error) error
}

// HookMiddlewareFunc 函数式中间件
type HookMiddlewareFunc struct {
	BeforeFn func(ctx context.Context, hctx *HookContext) error
	AfterFn  func(ctx context.Context, hctx *HookContext, err error) error
}

func (m *HookMiddlewareFunc) Before(ctx context.Context, hctx *HookContext) error {
	if m.BeforeFn != nil {
		return m.BeforeFn(ctx, hctx)
	}
	return nil
}

func (m *HookMiddlewareFunc) After(ctx context.Context, hctx *HookContext, err error) error {
	if m.AfterFn != nil {
		return m.AfterFn(ctx, hctx, err)
	}
	return err
}

// NewHookManager 创建钩子管理器
func NewHookManager() *HookManager {
	return &HookManager{
		hooks:      make(map[HookPoint][]Hook),
		stats:      NewHookStats(),
		middleware: make([]HookMiddleware, 0),
	}
}

func (m *HookManager) Register(point HookPoint, fn HookFunc) {
	m.RegisterWithPriority(point, fn, 0)
}

func (m *HookManager) RegisterWithPriority(point HookPoint, fn HookFunc, priority int) {
	m.RegisterConditional(point, fn, priority, Always, "")
}

func (m *HookManager) RegisterConditional(point HookPoint, fn HookFunc, priority int, condition HookCondition, id string) {
	m.RegisterConditionalInPhase(PhaseExecution, point, fn, priority, condition, id)
}

func (m *HookManager) RegisterInPhase(phase HookPhase, point HookPoint, fn HookFunc) {
	m.RegisterConditionalInPhase(phase, point, fn, 0, Always, "")
}

func (m *HookManager) refreshSnapshots() {
	newHooks := make(map[HookPoint][]Hook, len(m.hooks))
	for p, hs := range m.hooks {
		newHooks[p] = append([]Hook(nil), hs...)
	}
	m.hooksSnap.Store(&newHooks)

	newMid := append([]HookMiddleware(nil), m.middleware...)
	m.middlewareSnap.Store(&newMid)
}

func (m *HookManager) RegisterConditionalInPhase(phase HookPhase, point HookPoint, fn HookFunc, priority int, condition HookCondition, id string) {
	m.mu.Lock()
	defer m.mu.Unlock()

	hook := Hook{
		Point:     point,
		Func:      fn,
		Priority:  priority,
		Condition: condition,
		ID:        id,
		Phase:     phase,
	}
	m.hooks[point] = append(m.hooks[point], hook)

	for i := len(m.hooks[point]) - 1; i > 0; i-- {
		if m.hooks[point][i].Priority < m.hooks[point][i-1].Priority {
			m.hooks[point][i], m.hooks[point][i-1] = m.hooks[point][i-1], m.hooks[point][i]
		} else {
			break
		}
	}
	m.refreshSnapshots()
}

func (m *HookManager) Use(middleware HookMiddleware) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.middleware = append(m.middleware, middleware)
	m.refreshSnapshots()
}

func (m *HookManager) Fire(ctx context.Context, hctx *HookContext) error {
	var hooks []Hook
	if snap := m.hooksSnap.Load(); snap != nil {
		hooks = (*snap)[hctx.Point]
	} else {
		m.mu.RLock()
		hooks = append([]Hook(nil), m.hooks[hctx.Point]...)
		m.mu.RUnlock()
	}

	var mids []HookMiddleware
	if snap := m.middlewareSnap.Load(); snap != nil {
		mids = *snap
	} else {
		m.mu.RLock()
		mids = append([]HookMiddleware(nil), m.middleware...)
		m.mu.RUnlock()
	}

	var execErr error

	for _, mid := range mids {
		if err := mid.Before(ctx, hctx); err != nil {
			m.stats.Record(hctx.Point, err)
			return err
		}
	}

	for _, phase := range phaseOrder {
		for _, hook := range hooks {
			if hook.Phase != phase {
				continue
			}
			if hook.Condition == nil || hook.Condition(ctx, hctx) {
				if err := hook.Func(ctx, hctx); err != nil {
					execErr = err
					m.stats.Record(hctx.Point, err)
					break
				}
			}
		}
		if execErr != nil {
			break
		}
	}

	if execErr == nil {
		m.stats.Record(hctx.Point, nil)
	}

	for i := len(mids) - 1; i >= 0; i-- {
		afterErr := mids[i].After(ctx, hctx, execErr)
		if afterErr != nil {
			if afterErr == errRecovered {
				execErr = nil
			} else {
				execErr = afterErr
			}
		}
	}

	return execErr
}

func (m *HookManager) Remove(point HookPoint) {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.hooks, point)
	m.refreshSnapshots()
}

func (m *HookManager) RemoveByID(id string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	for point, hooks := range m.hooks {
		filtered := hooks[:0]
		for _, h := range hooks {
			if h.ID != id {
				filtered = append(filtered, h)
			}
		}
		m.hooks[point] = filtered
	}
	m.refreshSnapshots()
}

func (m *HookManager) Clear() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.hooks = make(map[HookPoint][]Hook)
	m.middleware = nil
	m.refreshSnapshots()
}

func (m *HookManager) Count(point HookPoint) int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return len(m.hooks[point])
}

func (m *HookManager) TotalCount() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	total := 0
	for _, hooks := range m.hooks {
		total += len(hooks)
	}
	return total
}

func (m *HookManager) Stats() *HookStats {
	return m.stats
}

func (m *HookManager) RegisteredPoints() []HookPoint {
	m.mu.RLock()
	defer m.mu.RUnlock()
	points := make([]HookPoint, 0, len(m.hooks))
	for p := range m.hooks {
		points = append(points, p)
	}
	return points
}

// ===== 便捷方法 =====

func (m *HookManager) OnReasoning(ctx context.Context, thought *core.Thought) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookAfterLLM
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnComplete(ctx context.Context, resp *core.Response) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookOnComplete
	hctx.Response = resp
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnTurnComplete(ctx context.Context, turn int, thought *core.Thought) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookAfterTurn
	hctx.Turn = turn
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnToolUse(ctx context.Context, tc *core.ToolCall) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookBeforeTool
	hctx.ToolCall = tc
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnToolResult(ctx context.Context, result *core.ToolResult) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookAfterTool
	hctx.ToolResult = result
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnError(ctx context.Context, err error) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookOnError
	hctx.Error = err
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnStream(ctx context.Context, event *core.StreamEvent) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookOnStream
	hctx.StreamChunk = event
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnStreamStart(ctx context.Context) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookOnStreamStart
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnStreamEnd(ctx context.Context, duration time.Duration) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookOnStreamEnd
	hctx.Duration = duration
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnStateChange(ctx context.Context, agentID, oldState, newState, reason string) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookOnStateChange
	hctx.AgentID = agentID
	hctx.OldState = oldState
	hctx.NewState = newState
	hctx.Reason = reason
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnMemoryRead(ctx context.Context, query string, result any) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookAfterMemoryRead
	hctx.MemoryQuery = query
	hctx.MemoryResult = result
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnMemoryWrite(ctx context.Context, sessionID string) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookAfterMemoryWrite
	hctx.SessionID = sessionID
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnContextWindowUpdate(ctx context.Context, usage float64, limit int) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookContextWindowUpdate
	hctx.ContextWindowUsage = usage
	hctx.ContextWindowLimit = limit
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnShutdown(ctx context.Context) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookBeforeShutdown
	_ = m.Fire(ctx, hctx)
}

func (m *HookManager) OnShutdownComplete(ctx context.Context) {
	hctx := AcquireHookContext()
	defer ReleaseHookContext(hctx)
	hctx.Point = HookAfterShutdown
	_ = m.Fire(ctx, hctx)
}

// AllHookPoints 返回所有定义的钩子点常量
func AllHookPoints() []HookPoint {
	return []HookPoint{
		HookBeforeRun, HookAfterRun,
		HookBeforeTurn, HookAfterTurn,
		HookBeforeLLM, HookAfterLLM,
		HookBeforeTool, HookAfterTool,
		HookOnError, HookOnComplete,
		HookBeforeRAG, HookAfterRAG,
		HookBeforePipelineStep, HookAfterPipelineStep,
		HookBeforeHandoff, HookAfterHandoff,
		HookBeforeParallelAgent, HookAfterParallelAgent,
		HookBeforeDAGNode, HookAfterDAGNode,
		HookOnStream, HookOnStreamStart, HookOnStreamEnd,
		HookBeforeMemoryRead, HookAfterMemoryRead,
		HookBeforeMemoryWrite, HookAfterMemoryWrite,
		HookContextWindowUpdate, HookContextWindowFull,
		HookBeforeToolParse, HookAfterToolParse,
		HookOnMetricsCollect,
		HookBeforeShutdown, HookAfterShutdown,
		HookOnStateChange,
	}
}

// HookPointCategory 钩子点分类
func HookPointCategory(p HookPoint) string {
	switch p {
	case HookBeforeRun, HookAfterRun, HookBeforeShutdown, HookAfterShutdown:
		return "lifecycle"
	case HookBeforeTurn, HookAfterTurn, HookOnComplete, HookOnError, HookOnStateChange:
		return "execution"
	case HookBeforeLLM, HookAfterLLM:
		return "llm"
	case HookBeforeTool, HookAfterTool, HookBeforeToolParse, HookAfterToolParse:
		return "tool"
	case HookBeforeRAG, HookAfterRAG, HookBeforeMemoryRead, HookAfterMemoryRead,
		HookBeforeMemoryWrite, HookAfterMemoryWrite:
		return "memory"
	case HookOnStream, HookOnStreamStart, HookOnStreamEnd:
		return "stream"
	case HookContextWindowUpdate, HookContextWindowFull:
		return "context"
	case HookBeforePipelineStep, HookAfterPipelineStep, HookBeforeHandoff, HookAfterHandoff,
		HookBeforeParallelAgent, HookAfterParallelAgent, HookBeforeDAGNode, HookAfterDAGNode:
		return "orchestration"
	case HookOnMetricsCollect:
		return "observability"
	default:
		return "unknown"
	}
}

// LoggingMiddleware 日志中间件
func LoggingMiddleware() *HookMiddlewareFunc {
	return &HookMiddlewareFunc{
		BeforeFn: func(_ context.Context, hctx *HookContext) error { return nil },
		AfterFn:  func(_ context.Context, hctx *HookContext, err error) error { return err },
	}
}

// MetricsCollectionMiddleware 指标收集中间件
func MetricsCollectionMiddleware(stats *HookStats) *HookMiddlewareFunc {
	return &HookMiddlewareFunc{
		BeforeFn: func(_ context.Context, _ *HookContext) error { return nil },
		AfterFn: func(_ context.Context, hctx *HookContext, err error) error {
			if stats != nil {
				stats.Record(hctx.Point, err)
			}
			return err
		},
	}
}

// TimeoutMiddleware 超时中间件
func TimeoutMiddleware(timeout time.Duration) *HookMiddlewareFunc {
	return &HookMiddlewareFunc{
		BeforeFn: func(ctx context.Context, hctx *HookContext) error { return nil },
		AfterFn: func(ctx context.Context, _ *HookContext, err error) error {
			select {
			case <-ctx.Done():
				return fmt.Errorf("hook execution cancelled")
			default:
				return err
			}
		},
	}
}

// errRecovered sentinel error
var errRecovered = errors.New("hook error recovered")

// phaseOrder Hook 执行阶段顺序
var phaseOrder = []HookPhase{PhaseValidation, PhasePreProcessing, PhaseExecution, PhasePostProcessing}

// ErrorRecoveryMiddleware 错误恢复中间件
func ErrorRecoveryMiddleware(onError func(HookPoint, error)) *HookMiddlewareFunc {
	return &HookMiddlewareFunc{
		BeforeFn: func(_ context.Context, _ *HookContext) error { return nil },
		AfterFn: func(ctx context.Context, hctx *HookContext, err error) error {
			if err != nil && onError != nil {
				onError(hctx.Point, err)
			}
			if err != nil {
				return errRecovered
			}
			return nil
		},
	}
}

File: sdk/typescript/src/agent/hooks.ts
/**
 * Agent 钩子系统 — 从 react-loop.ts 拆分
 *
 * 包含：StreamEvent、HookPoint、HookContext、HookFunc、HookManager
 *
 * 与 Go 端 hooks 包对齐，提供 Agent 生命周期钩子注册与触发机制。
 * HookManager 使用 ObjectPool 复用 HookContext，减少热路径 GC 压力。
 */
import type { Message, ToolCall, Response, ToolResult } from '../types.js';
import { ObjectPool } from '../jsonutil/pool.js';

// ===== 流式事件类型 =====

/** 流式事件联合类型，表示 Agent 运行过程中产生的各类事件。
 *
 * 事件类型：
 * - token: LLM 输出的文本片段（流式 token）
 * - tool_call: LLM 请求调用工具
 * - tool_result: 工具执行完成
 * - turn_end: 一个 ReAct 循环轮次结束
 * - done: Agent 运行完成，包含最终响应
 * - error: 运行过程中发生错误
 */
export type StreamEvent =
  | { type: 'token'; content: string }
  | { type: 'tool_call'; toolCall: ToolCall; turn: number }
  | { type: 'tool_result'; result: ToolResult; turn: number }
  | { type: 'turn_end'; turn: number }
  | { type: 'done'; response: Response }
  | { type: 'error'; error: Error };

/** 钩子触发点类型，覆盖 Agent 生命周期的关键节点。
 *
 * 与 Go 端 HookPoint 对齐，支持以下阶段：
 * - 运行阶段：before_run / after_run / on_complete / on_error
 * - 轮次阶段：before_turn / after_turn
 * - LLM 阶段：before_llm / after_llm
 * - 工具阶段：before_tool / after_tool / before_tool_parse / after_tool_parse
 * - 编排阶段：before_pipeline_step / after_pipeline_step / before_handoff / after_handoff
 * - 并行阶段：before_parallel_agent / after_parallel_agent / before_dag_node / after_dag_node
 * - 流式阶段：on_stream / on_stream_start / on_stream_end
 * - 记忆阶段：before_memory_read / after_memory_read / before_memory_write / after_memory_write
 * - 上下文窗口：context_window_update / context_window_full
 * - 指标阶段：on_metrics_collect
 * - 生命周期：before_shutdown / after_shutdown / on_state_change
 * - RAG 阶段：before_rag / after_rag
 */
export type HookPoint =
  | 'before_run'
  | 'after_run'
  | 'before_turn'
  | 'after_turn'
  | 'before_llm'
  | 'after_llm'
  | 'before_tool'
  | 'after_tool'
  | 'on_error'
  | 'on_complete'
  // Extended hook points (matching Go framework)
  | 'before_rag'
  | 'after_rag'
  | 'before_pipeline_step'
  | 'after_pipeline_step'
  | 'before_handoff'
  | 'after_handoff'
  | 'before_parallel_agent'
  | 'after_parallel_agent'
  | 'before_dag_node'
  | 'after_dag_node'
  | 'on_stream'
  | 'on_stream_start'
  | 'on_stream_end'
  | 'before_memory_read'
  | 'after_memory_read'
  | 'before_memory_write'
  | 'after_memory_write'
  | 'context_window_update'
  | 'context_window_full'
  | 'before_tool_parse'
  | 'after_tool_parse'
  | 'on_metrics_collect'
  | 'before_shutdown'
  | 'after_shutdown'
  | 'on_state_change';

/** 钩子上下文，包含 Agent 当前运行状态快照，传递给各钩子函数。
 *
 * 与 Go 端 HookContext 对齐，字段含义：
 * - agentID: Agent 标识
 * - sessionID: 会话标识
 * - point: 触发钩子的事件点
 * - turn: 当前轮次编号
 * - message: 当前消息（可选）
 * - response: 当前响应（可选）
 * - toolCall: 当前工具调用（可选）
 * - toolResult: 工具执行结果（可选）
 * - error: 错误信息（可选）
 * - metadata: 附加元数据（可选）
 * - requestID: 请求 ID，用于可观测性关联
 * - streamChunk: 流式数据块
 * - duration: 当前阶段耗时（毫秒）
 * - oldState: 状态变更前状态
 * - newState: 状态变更后状态
 * - reason: 变更原因
 * - memoryQuery: 记忆查询语句
 * - memoryResult: 记忆查询结果
 * - contextWindowUsage: 当前上下文窗口使用量
 * - contextWindowLimit: 上下文窗口上限
 */
export interface HookContext {
  agentID: string;
  sessionID: string;
  point: HookPoint;
  turn: number;
  message?: Message;
  response?: Response;
  toolCall?: ToolCall;
  toolResult?: ToolResult;
  error?: Error;
  metadata?: Record<string, unknown>;
  // Extended fields (matching Go framework)
  requestID?: string;
  streamChunk?: StreamEvent;
  duration?: number;
  oldState?: string;
  newState?: string;
  reason?: string;
  memoryQuery?: string;
  memoryResult?: unknown;
  contextWindowUsage?: number;
  contextWindowLimit?: number;
}

/** 钩子函数类型，接收 HookContext 并返回 void 或 Promise<void> */
export type HookFunc = (ctx: HookContext) => Promise<void> | void;

/** 钩子管理器，负责注册、触发和移除钩子函数。
 *
 * 使用方式：
 *   const hooks = new HookManager();
 *   hooks.register('before_run', (ctx) => { console.log(ctx.turn); });
 *   await hooks.fireHook({ agentID, sessionID, point: 'before_run', turn: 0 });
 */
export class HookManager {
  private hooks: Map<HookPoint, HookFunc[]> = new Map();
  /** HookContext 对象池，复用热路径对象减少 GC 压力（与 Go 端 sync.Pool 对齐） */
  private ctxPool = new ObjectPool<HookContext>(
    () => ({ agentID: '', sessionID: '', point: 'before_run', turn: 0 }),
    (ctx) => {
      ctx.agentID = '';
      ctx.sessionID = '';
      ctx.point = 'before_run';
      ctx.turn = 0;
      ctx.message = undefined;
      ctx.response = undefined;
      ctx.toolCall = undefined;
      ctx.toolResult = undefined;
      ctx.error = undefined;
      ctx.metadata = undefined;
      ctx.requestID = undefined;
      ctx.streamChunk = undefined;
      ctx.duration = undefined;
      ctx.oldState = undefined;
      ctx.newState = undefined;
      ctx.reason = undefined;
      ctx.memoryQuery = undefined;
      ctx.memoryResult = undefined;
      ctx.contextWindowUsage = undefined;
      ctx.contextWindowLimit = undefined;
    },
    64,
  );

  /** 注册钩子函数到指定触发点 */
  register(point: HookPoint, fn: HookFunc): void {
    if (!this.hooks.has(point)) {
      this.hooks.set(point, []);
    }
    this.hooks.get(point)!.push(fn);
  }

  /** 触发指定钩子点的所有注册函数（使用对象池减少 GC）
   *
   * 与 Go 端 fireHookWithPool 对齐，在 ReAct 热路径中复用 HookContext 对象。
   */
  async fireHook(point: HookPoint, opts: Omit<Partial<HookContext>, 'point'>): Promise<void> {
    // P0-6: 事件订阅者检查 — 无订阅者时跳过 payload 构造，避免热路径上的无用开销
    if (!this.hasSubscriber(point)) return;
    const ctx = this.ctxPool.get();
    ctx.point = point;
    if (opts.agentID !== undefined) ctx.agentID = opts.agentID;
    if (opts.sessionID !== undefined) ctx.sessionID = opts.sessionID;
    if (opts.turn !== undefined) ctx.turn = opts.turn;
    if (opts.message !== undefined) ctx.message = opts.message;
    if (opts.response !== undefined) ctx.response = opts.response;
    if (opts.toolCall !== undefined) ctx.toolCall = opts.toolCall;
    if (opts.toolResult !== undefined) ctx.toolResult = opts.toolResult;
    if (opts.error !== undefined) ctx.error = opts.error;
    if (opts.metadata !== undefined) ctx.metadata = opts.metadata;
    if (opts.requestID !== undefined) ctx.requestID = opts.requestID;
    if (opts.streamChunk !== undefined) ctx.streamChunk = opts.streamChunk;
    if (opts.duration !== undefined) ctx.duration = opts.duration;
    if (opts.oldState !== undefined) ctx.oldState = opts.oldState;
    if (opts.newState !== undefined) ctx.newState = opts.newState;
    if (opts.reason !== undefined) ctx.reason = opts.reason;
    if (opts.memoryQuery !== undefined) ctx.memoryQuery = opts.memoryQuery;
    if (opts.memoryResult !== undefined) ctx.memoryResult = opts.memoryResult;
    if (opts.contextWindowUsage !== undefined) ctx.contextWindowUsage = opts.contextWindowUsage;
    if (opts.contextWindowLimit !== undefined) ctx.contextWindowLimit = opts.contextWindowLimit;
    try {
      await this.fire(ctx);
    } finally {
      this.ctxPool.put(ctx);
    }
  }

  /** 触发指定钩子点的所有注册函数 */
  async fire(ctx: HookContext): Promise<void> {
    const fns = this.hooks.get(ctx.point) ?? [];
    for (const fn of fns) {
      await fn(ctx);
    }
  }

  /** P0-6: 检查指定钩子点是否有订阅者（与 Go 端 hasEventSubscriber 对齐） */
  hasSubscriber(point: HookPoint): boolean {
    return (this.hooks.get(point)?.length ?? 0) > 0;
  }

  /** 移除指定钩子点的所有注册函数 */
  remove(point: HookPoint): void {
    this.hooks.delete(point);
  }

  /** 清空所有钩子注册 */
  clear(): void {
    this.hooks.clear();
  }

  /** 查询指定钩子点的注册函数数量 */
  count(point: HookPoint): number {
    return this.hooks.get(point)?.length ?? 0;
  }
}
